### ISSUE TITLE
[$200 bounty] formatter: experimentalSortImports.customGroups

### ISSUE THREAD

**Issue by MatissJanis** [oxc-project/oxc#17076]
Objective: implement [customGroups](https://perfectionist.dev/rules/sort-imports.html#customgroups) configuration option to the `experimentalSortImports`.

Testing: feel free to use [ActualBudget](https://github.com/actualbudget/actual). We already use the eslint plugin for sorting imports, so just moving it over to `oxfmt.json` and running `yarn oxfmt` should work and yield no changed files after `customGroups` is implemented.

---

**Bounty** - $200

We (ActualBudget) are sponsoring this. Once the `customGroups` config option has been implemented and we have verified that it works as expected in ActualBudget repository (see testing section above) - you will be able to submit an invoice to our [open collective page](https://opencollective.com/actual). We will make sure to approve it. On AB side we have already approved and budgeted this expense.

**Comment by leaysgur** [comment]
Advice for those interested in this issue:

- Please read and follow our guideline, especially about using AI
  - https://github.com/oxc-project/oxc/blob/main/CONTRIBUTING.md
- `eslint-plugin-perfectionist` is highly customizable, and as a result, complex
  - You need to understand what can be done with `groups` and `customGroups`
    - https://perfectionist.dev/rules/sort-imports#groups
- Please share your plan first, rather than send a PR
  - I don't feel confident I could review this feature even if a massive PR were submitted
- Also, I recommend just to start by adding minimum set of feature
  - e.g. only supports this shape: `{ groupName: 'myFoo', elementNamePattern: '@myfoo' }`
    - No `regex`, just check as prefix
  - (This was my original plan with no ETA, tho)

**Comment by nilptr** [comment]
I'd like to try it.

**Comment by nilptr** [comment]
The request is to add the support for [customGroups](https://perfectionist.dev/rules/sort-imports.html#customgroups) option, so I assume I just need to add the new option and change the [into_match_group_idx](https://github.com/oxc-project/oxc/blob/main/crates/oxc_formatter/src/ir_transform/sort_imports/compute_metadata.rs#L108) logic.

The new option will be like

```rust
struct CustomGroupRule {
    selector: Option<ImportSelector>,
    modifiers: Option<Vec<ImportModifier>>,
    element_name_pattern: Option<String>,
}

struct CustomGroupDefinition {
    group_name: String,
    any_of: Vec<CustomGroupRule>,
}

pub struct SortImportsOptions {
    // ...other options

    // won't change
    pub groups: Vec<Vec<String>>,
    // New
    pub customGroups: Option<Vec<CustomGroupDefinition>>,
}
```

Perfectionist used a very inefficient way to determine the group of an import.

1. It gets the `{ selectors: []; modifiers: [] }` of an import (both `selectors` and `modifiers` are sorted by importance)
2. Then it enumerates the import's all possible groups `predefinedGroups` (all permutations of modifiers + single selector)
3. Then it checks if the import matches any custom group
4. Then it finds the first group in `predefinedGroups` that is also in the `groups`

I feel we can sort the non-custom groups first, and match `{ selectors: []; modifiers: [] }` against the sorted non-custom group list.

**Comment by leaysgur** [comment]
> Perfectionist used a very inefficient way to determine the group of an import.
> I feel we can sort the non-custom groups first,

I'm open to changes, especially if they simplify things. 👌🏻 

However, I also think there's a reason it's set up this way, so we should verify whether it truly poses no issues.

**Comment by nilptr** [comment]
> so we should verify whether it truly poses no issues.

Regarding this concern, I have forked eslint-plugin-perfectionist and changed to [my algorithm](https://github.com/azat-io/eslint-plugin-perfectionist/compare/main...nilptr:eslint-plugin-perfectionist:nilptr/test/simplify-sort-imports-algorithm) in `nilptr/test/simplify-sort-imports-algorithm` branch. It can pass all existing unit tests. 

<img width="1824" height="1488" alt="Image" src="https://github.com/user-attachments/assets/c289b884-0317-4e2b-8dbc-07bd6d66cc93" />

As for oxfmt, it's still in progress. I can only work on this after work, please give me a little more time.

**Comment by leaysgur** [comment]
Great work @nilptr 👏🏻 #17576 has merged and will be released as v0.25.

The docs will be updated soon, but here is an example usage.

```json
  "experimentalSortImports": {
    // Define your group names
    "customGroups": [
      {
        "groupName": "react", // This will be used in `groups`
        "elementNamePattern": ["react"] // Array of prefixes, not regex
      },
      {
        "groupName": "our-core",
        "elementNamePattern": ["our-core/"]
      },
    ],
    "groups": [
      "side-effect",
      "react", // <-
      "builtin",
      "external",
      "our-core", // <-
      ["parent", "sibling", "index"],
      "type",
    ],
  },
```

**Comment by jycouet** [comment]
Very nice 🚀

If possible, in the example can you add
- How to have gaps between groups ?
- what are all keys (parent, ...)

Maybe it will be part of docs ☺️
Can't wait to use it, well done 💪

**Comment by oleasteo** [comment]
Great work!

I know it's not part of the bounty as the OP example suffices with prefix patterns, but is there any follow-up planned to support suffix checks as well?

For context, my use-case would be to enforce `*.mock.ts` at the very beginning of files as those contain my `vi.mock` calls (I dislike the \_\_mocks\_\_/ pattern).

Another valid use-case for suffix checks could be to group fixture imports (e.g., `*.json` or `*.xml` and such).

**Comment by vlechemin** [comment]
Hi,

I tried with `0.25.0` this config file:

```json
{
    "experimentalSortImports": {
        "customGroups": [
            {
                "elementNamePattern": ["~/validators/"],
                "groupName": "validators"
            },
            {
                "elementNamePattern": ["~/composable/"],
                "groupName": "composable"
            },
            {
                "elementNamePattern": ["~/components/"],
                "groupName": "components"
            },
            {
                "elementNamePattern": ["~/services/"],
                "groupName": "services"
            },
            {
                "elementNamePattern": ["~/widgets/"],
                "groupName": "widgets"
            },
            {
                "elementNamePattern": ["~/stores/"],
                "groupName": "stores"
            },
            {
                "elementNamePattern": ["~/logics/"],
                "groupName": "logics"
            },
            {
                "elementNamePattern": ["~/assets/"],
                "groupName": "assets"
            },
            {
                "elementNamePattern": ["~/utils/"],
                "groupName": "utils"
            },
            {
                "elementNamePattern": ["~/pages/"],
                "groupName": "pages"
            },
            {
                "elementNamePattern": ["~/ui/"],
                "groupName": "ui"
            }
        ],
        "groups": [
            ["builtin", "external"],
            "internal",
            "stores",
            "services",
            "validators",
            "utils",
            "logics",
            "composable",
            "ui",
            "components",
            "pages",
            "widgets",
            "assets",
            "parent",
            "sibling",
            "side-effect",
            "index",
            "style",
            "unknown"
        ]
    }
}
```

It does not reproduce the code in the test, but this instead:

```ts
import ComponentA from "~/components/ComponentA.vue";
import ComponentB from "~/components/ComponentB.vue";
import ComponentC from "~/components/ComponentC.vue";
import { useDebounce } from "~/composable/useDebounce.ts";
import { useFetch } from "~/composable/useFetch.ts";
import { useMouseMove } from "~/composable/useMouseMove.ts";
import { getCart } from "~/services/cartService.ts";
import { useCartStore } from "~/stores/cartStore.ts";
import { useUserStore } from "~/stores/userStore.ts";
import { formattingDate } from "~/utils/dateTime.ts";
import { connect } from "~/utils/ws.ts";

import CartComponentA from "./cart/CartComponentA.vue";
import CartComponentB from "./cart/CartComponentB.vue";
```

Am I missing something? Thanks

**Comment by leaysgur** [comment]
> - How to have gaps between groups?

Currently, the only option for spacing between groups is the global `newlinesBetween` option.

https://oxc.rs/docs/guide/usage/formatter/config-file-reference.html#experimentalsortimports-newlinesbetween

In the original implementation, this can be specified individually on the group side as well, but we have not implemented this yet. 👉🏻 #17830

> - what are all keys (parent, ...)

For now, please refer to this.

https://github.com/oxc-project/oxc/blob/00ff75f258bc388bc26b2e7d2cfbd94e17a5a93f/apps/oxfmt/src/core/oxfmtrc.rs#L564-L615

Documentation will be updated soon.

---

> I know it's not part of the bounty as the OP example suffices with prefix patterns, but is there any follow-up planned to support suffix checks as well?

Yes. Ultimately, I'm thinking to enable regular expressions (or similar expression), just like the original.

However,

- Using regular expressions in Rust requires an external library
- The syntax differs from JavaScript's regular expressions
- If we implement it, I'd also like to update other parts that should support regular expressions

For these reasons, I've omitted it for now.

If this is an important, please create a separate issue. We'd like to confirm the demand before implementing it.

---

> Am I missing something? Thanks

Sorry, there was a bug in 0.25 that prevented it from working when used via the CLI.
We plan to release 0.26 as extra release.

**Comment by stunaz** [comment]
doesn't seem to work for me.. in a monorepo having a group with prefix `@acme/` does not group it... still categorized as external

**Comment by vlechemin** [comment]
> doesn't seem to work for me.. in a monorepo having a group with prefix `@acme/` does not group it... still categorized as external

It will be fixed in 0.26

**Comment by leaysgur** [comment]
@nilptr Thanks again and I'm going to close this issue now.

I believe the original requirements have been [met](https://github.com/actualbudget/actual/pull/6719) and ready to receive the bounty, so please proceed with the necessary steps. 💸

**Comment by vlechemin** [comment]
~~Hi @leaysgur, I'm trying to make it work in Vue files, I have `embeddedLanguageFormatting` to `auto`, but it doesn't seem to be enough, is it supported? Thanks~~

(tracked in https://github.com/oxc-project/oxc/issues/16608)

**Comment by MatissJanis** [comment]
Thanks @nilptr for implementing this! Feel free to submit an invoice on our [OpenCollective](https://opencollective.com/actual) page and I'll be happy to approve it.

Select "submit expense" -> "invoice".

Please let me know when this is done by posting here again.

**Comment by nilptr** [comment]
Thanks @MatissJanis. I have submitted an invoice [OpenCollective expense #282054](https://opencollective.com/actual/expenses/282054).

**Comment by MatissJanis** [comment]
Approved! It takes a few days to process. Should arrive to your bank in 2 weeks time if not sooner.
